qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,754 @@
+    generated until you assign all unbounded parameters.
+    Note that parameters will be assigned through the schedule playing the pulse.
+
+
+    .. _symbolic_pulse_serialize:
+
+    .. rubric:: Serialization
+
+    The :class:`~SymbolicPulse` subclass is QPY serialized with symbolic expressions.
+    A user can therefore create a custom pulse subclass with a novel envelope and constraints,
+    and then one can instantiate the class with certain parameters to run on a backend.
+    This pulse instance can be saved in the QPY binary, which can be loaded afterwards
+    even within the environment not having original class definition loaded.
+    This mechanism allows us to easily share a pulse program including custom pulse instructions
+    with collaborators or to directly submit the program to the quantum computer backend
+    in the parametric form (i.e. not sample data). This greatly reduces amount of data transferred.
+    The waveform sample data to be loaded in the control electronics
+    will be reconstructed with parameters and deserialized expression objects by the backend.
+
+    .. note::
+
+        Currently QPY serialization of :class:`SymbolicPulse` is not available.
+        This feature will be implemented shortly.
+        Note that data transmission in the parametric form requires your quantum computer backend
+        to support QPY framework with :class:`SymbolicPulse` available. Otherwise, the pulse data
+        might be converted into :class:`Waveform`.
+    """
+
+    # Lambdify caches keyed on sympy expressions. Returns the corresponding callable.
+    _callable_envelope = LamdifiedExpression(
+        common_params=["t", "duration", "amp"],
+        attribute="_envelope_expr",
+    )
+    _callable_consts = LamdifiedExpression(
+        common_params=["duration", "amp"],
+        attribute="_consts_expr",
+    )
+
+    def __init__(
+        self,
+        pulse_type: str,
+        parameters: Dict[str, Union[ParameterExpression, complex]],
+        name: Optional[str] = None,
+        limit_amplitude: Optional[bool] = None,
+    ):
+        """Create a parametric pulse and validate the input parameters.
+
+        Args:
+            pulse_type: Display name of this pulse shape.
+            parameters: Dictionary of pulse parameters. This must include "duration".
+            name: Display name for this particular pulse envelope.
+            limit_amplitude: If ``True``, then limit the absolute value of the amplitude of the
+                waveform to 1. The default is ``True`` and the amplitude is constrained to 1.
+
+        Raises:
+            PulseError: When not all parameters are listed in the attribute :attr:`PARAM_DEF`.
+        """
+        if "duration" not in parameters:
+            raise PulseError("'duration' is not defined for this pulse.")
+        if "amp" not in parameters:
+            raise PulseError("'amp' is not defined for this pulse.")
```

Comment:
In https://github.com/Qiskit/qiskit-terra/pull/7821/commits/d9574586f0085523c123b28340d2a56356deec0b for symengine I treated `amp` as a special (required) parameter as a _scaling factor_ of the envelope, so all envelopes must be defined in the form of `amp * F(t)`. This reduces the complexity because, symengine doesn't allow lamdify args which is mixture of complex and float (i.e. if you pass complex, then everything become complex). 

> I ask partly because I know someone was implementing a pulse to propose as a new parametric pulse supported by the IBM provider and it would not have an amp.

Probably we can make `amp` default to 1.0, but this parameter will appear in `.parameters`.